### PR TITLE
Cankov/ios debug

### PR DESCRIPTION
**Pull request by PanayotCankov**
Implemented on device debugging for iOS

Build, deploy and start the app on device, launch the front end tools in Safai, attach the debugger through USB as soon as possible. In future we will try to implement in the iOS runtime a brake before the first statement:
```tns debug ios --debug-brk```

Start Safari with the front end tools, and try to attach to an already running app:
```tns debug ios --start```

### CODE REVIEW

File: mobile/ios/ios-proxy-services.ts [hunk 1]
```diff
@@ -10,6 +10,11 @@ import util = require("util");
 import helpers = require("./../../helpers");
 import net = require("net");
 import MobileHelpers = require("./../mobile-helper");
+import bufferpack = require("bufferpack");
+import plistlib = require("plistlib");
+
+// This should stay in ios-core...
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: Which one

File: mobile/ios/ios-device.ts [hunk 2]
```diff
@@ -14,10 +15,17 @@ import helpers = require("./../../helpers");
 import hostInfo = require("./../../host-info");
 import options = require("./../../options");
 
+import options = require("./../../options");
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]: options is already required 2 rows above

File: mobile/ios/ios-device.ts [hunk 3]
```diff
@@ -320,11 +333,135 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+        var tnsIosPackage = this.$npm.install("tns-ios").wait();
+        var safariPath = path.join(tnsIosPackage, "WebInspectorUI/Safari/Main.html");
+        this.$childProcess.exec("open -a Safari " + safariPath).wait();
+    }
+
+    private attachWhenReady(packageName: string,
+        notificationProxyClient: iOSProxyServices.NotificationProxyClient): void {
+
+        var identifier = this.getIdentifier();
+        console.log("Device Identifier: " + identifier);
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]:
```Typescript
this.$logger.out("Device identifies %s", identifier);
```

File: mobile/ios/ios-device.ts [hunk 4]
```diff
@@ -320,11 +333,135 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+
+        var appReadyForAttachMessage = packageName + ":NativeScript.Debug.ReadyForAttach";
+        notificationProxyClient.addObserver(appReadyForAttachMessage, () => {
+            console.log("Got ReadyForAttach");
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]: Use logger if it's possible

File: mobile/ios/ios-device.ts [hunk 5]
```diff
@@ -320,11 +333,135 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+
+			var port = 8080;
+            var localServer = net.createServer((localSocket) => {
+                console.log("Front end connected to localhost");
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]: Use logger if possible

File: mobile/ios/ios-device.ts [hunk 6]
```diff
@@ -320,11 +333,135 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+			});
+
+            localServer.listen(port, () => {
+                console.log("Opened localhost 8080");
+            });
+        });
+    }
+
+    private printHowToTerminate() {
+        console.log("Setting up debugger proxy...");
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]: logger?

File: mobile/ios/ios-device.ts [hunk 7]
```diff
@@ -320,11 +333,135 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+    }
+
+    private detachDebugger(packageName: string): void {
+        console.log("detachDebugger: " + packageName);
+    }
+
+	private startAppWithDebugger(packageFile: string, packageName: string): void {
+        this.deploy(packageFile, packageName).wait();
+        this.attachAtStartup(packageName);
+        this.runApplication(packageName).wait();
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]: Is it a problem to call runApplication when attachAtStartup's addObserver is not finished (when we have not received  "Got AppLaunching")?

File: mobile/ios/ios-device.ts [hunk 8]
```diff
@@ -320,11 +333,135 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
 	}
 
     public debug(packageFile: string, packageName: string): IFuture<void> {
         return (() => {
-            this.$errors.fail({formatStr:"this will come in a future version", suppressCommandHelp: true});
+            if (options["get-port"]) {
+                this.$errors.fail({ formatStr: "this future is not supported in the selected platform", suppressCommandHelp: true });
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]:
```Typescript
this.$errors.failWithoutHelp("This future is not supported in the selected platform");
```

File: mobile/ios/ios-device.ts [hunk 9]
```diff
@@ -320,11 +333,135 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+            } else if (options["start"]) {
+                this.attachDebugger(packageName);
+            } else if (options["stop"]) {
+                this.$errors.fail({ formatStr: "this future is not supported in the selected platform", suppressCommandHelp: true });
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]:
```Typescript
this.$errors.failWithoutHelp("This future is not supported in the selected platform");
```

File: mobile/ios/ios-device.ts [hunk 10]
```diff
@@ -385,5 +521,25 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+			var socketValue = socketRef.deref();
+
+			return new net.Socket({ fd: socketValue });
+		}
+
+		return null;
+	}
+
+	private htons(port: number): number {
+		var result =  (port & 0xff00) >> 8 | (port & 0x00ff) << 8;
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]: awesome :joy_cat:

File: mobile/ios/ios-emulator-services.ts [hunk 11]
```diff
@@ -37,7 +39,43 @@ class IosEmulatorServices implements Mobile.IEmulatorPlatformServices {
 			this.startEmulatorCore(app, emulatorOptions);
 
 		}).future<void>()();
-	}
+    }
+
+    startAppWithDebugger(app: string, emulatorOptions?: Mobile.IEmulatorOptions) {
+        var tnsIosPackage = this.$npm.install("tns-ios", ["g"]).wait();
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-emulator-services.ts]: Is it possible to require sudo permissions here?
If node is installed with sudo, -g option may need sudo (or I'm totally incorrect).

File: mobile/ios/ios-emulator-services.ts [hunk 12]
```diff
@@ -37,7 +39,43 @@ class IosEmulatorServices implements Mobile.IEmulatorPlatformServices {
...
+        var safariPath = path.join(tnsIosPackage, "WebInspectorUI/Safari/Main.html");
+        this.$childProcess.exec("open -a Safari " + safariPath).wait();
+
+        this.startEmulator(app, emulatorOptions).wait();
+
+        // Start pumping "WaitForDebugger" messages to all Booted devices...
+
+        var deviceList = this.$childProcess.exec("xcrun simcrl list");
+        deviceList.
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-emulator-services.ts]: unfinished statement

File: mobile/ios/ios-emulator-services.ts [hunk 13]
```diff
@@ -37,7 +39,43 @@ class IosEmulatorServices implements Mobile.IEmulatorPlatformServices {
...
+
+        console.log("App started!");
+    }
+
+    attachDebugger(app: string, emulatorOptions?: Mobile.IEmulatorOptions) {
+
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-emulator-services.ts]: empty method?

File: mobile/ios/ios-emulator-services.ts [hunk 14]
```diff
@@ -37,7 +39,43 @@ class IosEmulatorServices implements Mobile.IEmulatorPlatformServices {
...
+        // xcrun simctl notify_post device message
+    }
+
+    debugEmulator(app: string, emulatorOptions?: Mobile.IEmulatorOptions): IFuture<void> {
+        return (() => {
+            if (options["get-port"]) {
+                this.$errors.fail({ formatStr: "this future is not supported in the selected platform", suppressCommandHelp: true });
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-emulator-services.ts]:
```Typescript
this.$errors.failWithoutHelp("This future is not supported in the selected platform");
```

File: mobile/ios/ios-emulator-services.ts [hunk 15]
```diff
@@ -37,7 +39,43 @@ class IosEmulatorServices implements Mobile.IEmulatorPlatformServices {
...
+            } else if (options["start"]) {
+                this.attachDebugger(app);
+            } else if (options["stop"]) {
+                this.$errors.fail({ formatStr: "this future is not supported in the selected platform", suppressCommandHelp: true });
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-emulator-services.ts]:
```Typescript
this.$errors.failWithoutHelp("This future is not supported in the selected platform");
```

File: mobile/ios/ios-proxy-services.ts [hunk 16]
```diff
@@ -220,25 +225,116 @@ export class InstallationProxyClient {
...
 			"ClientOptions": ""
 		});
-
 		return result;
-	}
+    }
+
+    public addObserver(name: string, callback: (name: string) => void) {
+        this.openSocket();
+        // console.log("ADD OBSERVER: " + name);
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: $logger or remove the line

File: mobile/ios/ios-proxy-services.ts [hunk 17]
```diff
@@ -220,25 +225,116 @@ export class InstallationProxyClient {
...
+        if (!this._plistService) {
+            this._plistService = this.$injector.resolve(iOSCore.PlistService, { service: this.device.startService(MobileServices.NOTIFICATION_PROXY), format: iOSCore.CoreTypes.kCFPropertyListBinaryFormat_v1_0 });
+            this._plistService.readSystemLog(this.handleData.bind(this));
+        }
+    }
+
+    private handleData(data: NodeBuffer): void {
+
+        var str = data.toString();
+        // console.log("Notification Data: " + str);
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: $logger or remove the line

File: mobile/ios/ios-proxy-services.ts [hunk 18]
```diff
@@ -220,25 +225,116 @@ export class InstallationProxyClient {
...
+            this.buffer = this.buffer.substr(end);
+
+            try {
+
+                //console.log("PLIST: " + plist);
+
+                plistlib.loadString(plist, (err: any, plist: any) => {
+                    if (err) {
+                        // ERROR
+                        //console.log("error: " + err);
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: $logger or $erros.fail

File: mobile/ios/ios-proxy-services.ts [hunk 19]
```diff
@@ -297,3 +393,228 @@ export class IOSSyslog {
...
+export class UsbmuxDeviceSocket {
+    private socket: net.NodeSocket;
+    private $coreFoundation;
+    private connected: boolean = false;
+
+    constructor(private usbmuxDevice: UsbmuxDevice, private port: number, private onSuccess: (socket: net.NodeSocket) => void) {
+        this.$coreFoundation = $injector.resolve(iOSCore.CoreFoundation);
+
+        if (hostInfo.isWindows()) {
+            // TODO: Find the port in graphite and use it...
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: Graphite? why?

File: mobile/ios/ios-proxy-services.ts [hunk 20]
```diff
@@ -297,3 +393,228 @@ export class IOSSyslog {
...
+        console.log("UsbmuxDeviceSocket Error");
+    }
+
+    private sendMessage(message: { [key: string]: {} }) {
+        // This does not seem to work with numbers: var data = this.$coreFoundation.dictToPlistEncoding(message, iOSCore.CoreTypes.kCFPropertyListXMLFormat_v1_0);
+        var data = plistlib.toString(message);
+
+        var version = 1;
+        var messageType = 8; // MessageType.PList = 8
+        var tag = 1; // I have no idea...
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: I love this comment :joy:

File: mobile/ios/ios-proxy-services.ts [hunk 21]
```diff
@@ -297,3 +393,228 @@ export class IOSSyslog {
...
+        this.socket.on("error", this.handleError.bind(this));
+    }
+
+    /**
+     * The callback will be invoked only once if the device with that serial number is connected
+     * or when the device with that serial number connects.
+     */
+    public withDevice(serialNumber: string, callback: (UsbmuxDevice) => void) {
+        for (var i = 0; i < this.devices.length; i++) {
+            if (this.devices[i].serialNumber == serialNumber) {
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: ===

File: mobile/ios/ios-proxy-services.ts [hunk 22]
```diff
@@ -297,3 +393,228 @@ export class IOSSyslog {
...
+
+            try {
+
+                // console.log("PLIST: " + plist);
+
+                plistlib.loadString(plist, (err: any, plist: any) => {
+                    try {
+                        var msgType = plist.value.MessageType.value;
+                        // console.log("UsbmuxDeviceListener: " + msgType);
+                        if (msgType == "Result") {
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: ``` === ```

File: mobile/ios/ios-proxy-services.ts [hunk 23]
```diff
@@ -297,3 +393,228 @@ export class IOSSyslog {
...
+                            if (plist.value.Number.value == 0) {
+                                // console.log("UsbmuxDeviceListener Success!");
+                            } else {
+                                // console.log("UsbmuxDeviceListener Failure!");
+                            }
+                        } else if (msgType == "Attached") {
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: ```===```

File: mobile/ios/ios-proxy-services.ts [hunk 24]
```diff
@@ -297,3 +393,228 @@ export class IOSSyslog {
...
+                            var usbmuxDevice = new UsbmuxDevice();
+                            usbmuxDevice.id = Number(plist.value.Properties.value.DeviceID.value);
+                            usbmuxDevice.serialNumber = plist.value.Properties.value.SerialNumber.value;
+                            this.Attached(usbmuxDevice);
+                        } else if (msgType == "Detached") {
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-proxy-services.ts]: ```===```

**Comment by rosen-vladimirov** [comment]
There are a lot of commented console.log calls, we should delete them or replace them with $logger.

File: mobile/ios/ios-device.ts (hunk 8 above)
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: future? maybe "feature"?

File: mobile/ios/ios-device.ts [hunk 25]
```diff
@@ -30,7 +32,8 @@ export class IOSDevice implements Mobile.IIOSDevice {
 		private $fs: IFileSystem,
 		private $injector: IInjector,
 		private $logger: ILogger,
-		private $mobileDevice: Mobile.IMobileDevice) {
+        private $mobileDevice: Mobile.IMobileDevice,
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: Mixed spaces and tabs (I know, it is silly)

File: mobile/ios/ios-device.ts [hunk 26]
```diff
@@ -303,6 +306,10 @@ export class IOSDevice implements Mobile.IIOSDevice {
 		}).future<void>()();
 	}
 
+	private getInterfaceType(): any {
+		return this.$mobileDevice.deviceGetInterfaceType(this.devicePointer);
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: deviceGetInterfaceType returns a number, should getInterfaceType return number as well?

File: mobile/ios/ios-device.ts [hunk 27]
```diff
@@ -320,11 +327,130 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+    private postDebugNotification(packageName: string, notification: string, notificationProxyClient: iOSProxyServices.NotificationProxyClient): void {
+        var attachRequestMessage = packageName + ":NativeScript.Debug.AttachRequest";
+        notificationProxyClient.postNotification(attachRequestMessage);
+        this.$logger.info("Send " + notification);
+    }
+
+    private attachAtStartup(packageName: string) {
+        this.openSafariFrontEnd(packageName);
+
+        var _injector = this.$injector;
```
Comment (Fatme) [inline mobile/ios/ios-device.ts]: You don't need this.

File: mobile/ios/ios-device.ts [hunk 28]
```diff
@@ -320,11 +327,130 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+
+        var identifier = this.getIdentifier();
+        this.$logger.info("Device Identifier: " + identifier);
+
+        var appReadyForAttachMessage = packageName + ":NativeScript.Debug.ReadyForAttach";
+        notificationProxyClient.addObserver(appReadyForAttachMessage, () => {
+            this.$logger.info("Got ReadyForAttach");
+
+            var devicePort = 8080;
+            var localPort = 8080;
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: both devicePort and localPorts are constants. Consider declaring them private static DEVICE_PORT members, to signify they are nor mutable.

File: mobile/ios/ios-device.ts [hunk 29]
```diff
@@ -320,11 +327,130 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+    }
+
+    private openSafariFrontEnd(packageName: string): void {
+        var tnsIosPackage = this.$npm.install("tns-ios").wait();
+        var safariPath = path.join(tnsIosPackage, "WebInspectorUI/Safari/Main.html");
+        this.$childProcess.exec("open -a Safari " + safariPath).wait();
+    }
+
+    private attachWhenReady(packageName: string,
+        notificationProxyClient: iOSProxyServices.NotificationProxyClient): void {
```
Comment (Fatme) [inline mobile/ios/ios-device.ts]: Do you need to pass `notificationProxyClient` as a parameter. I think you should be able to create it constructor.

File: mobile/ios/ios-device.ts [hunk 30]
```diff
@@ -320,11 +327,130 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+            this.attachWhenReady(packageName, notificationProxyClient);
+            this.postDebugNotification(packageName, "WaitForDebugger", notificationProxyClient);
+        });
+    }
+
+    private attachAtRuntime(packageName: string): void {
+        this.openSafariFrontEnd(packageName);
+
+        var _injector = this.$injector;
+        var notificationProxyClient: iOSProxyServices.NotificationProxyClient = _injector.resolve(iOSProxyServices.NotificationProxyClient, { device: this });
```
Comment (Fatme) [inline mobile/ios/ios-device.ts]: Just `this.$injector.resolve`

File: mobile/ios/ios-device.ts [hunk 31]
```diff
@@ -320,11 +327,130 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+                });
+                localSocket.on('end', () => {
+                    this.$logger.info('Localhost client disconnected!');
+                    process.exit(0);
+                });
+
+                deviceSocket = this.connectToPort(devicePort);
+                this.$logger.info("Connected localhost " + localPort + " to device " + devicePort);
+
+				buff.forEach((data) => {
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: Mixed tabs / spaces

File: mobile/ios/ios-device.ts [hunk 32]
```diff
@@ -320,11 +327,130 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+					deviceSocket.write(data);
+				});
+
+				deviceSocket.on('data', (data: NodeBuffer) => {
+					localSocket.write(data);
+				});
+
+				deviceSocket.on('end', (data: NodeBuffer) => {
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: You may omit the unused lambda argument

File: mobile/ios/ios-proxy-services.ts [hunk 33]
```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
...
+            if (index >= 0 && index < array.length) {
+                array.splice(index, 1);
+            }
+        }
+    }
+
+    private openSocket() {
+        if (!this.plistService) {
+            this.plistService = this.$injector.resolve(iOSCore.PlistService, { service: this.device.startService(MobileServices.NOTIFICATION_PROXY), format: iOSCore.CoreTypes.kCFPropertyListBinaryFormat_v1_0 });
+            this.plistService.readSystemLog(this.handleData.bind(this));
```
Comment (Fatme) [inline mobile/ios/ios-proxy-services.ts]: Do you need to read system log here?

File: mobile/ios/ios-device.ts (hunk 27 above)
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: Is this var needed?

File: mobile/ios/ios-proxy-services.ts [hunk 34]
```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
...
+        }
+    }
+
+    private handleData(data: NodeBuffer): void {
+        this.buffer += data.toString();
+
+        var start = this.buffer.indexOf("<plist");
+        var end = this.buffer.indexOf("</plist>") + 8;
```
Comment (Fatme) [inline mobile/ios/ios-proxy-services.ts]: Magic number! :smile_cat:  It seems that `8` is the length of `</plist>`. I suggest  you to extract it to separate variable

File: mobile/ios/ios-device.ts [hunk 35]
```diff
@@ -320,11 +327,130 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+        notificationProxyClient.addObserver(appLaunchMessage, () => {
+            this.$logger.info("Got AppLaunching");
+
+            this.attachWhenReady(packageName, notificationProxyClient);
+            this.postDebugNotification(packageName, "WaitForDebugger", notificationProxyClient);
+        });
+    }
+
+    private attachAtRuntime(packageName: string): void {
+        this.openSafariFrontEnd(packageName);
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: This function is very similar to attachAtStartup, can we refactor the common parts?

File: mobile/ios/ios-proxy-services.ts [hunk 36]
```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
...
+            var plist = this.buffer.substr(start, end);
+            this.buffer = this.buffer.substr(end);
+
+            try {
+                plistlib.loadString(plist, (err: any, plist: any) => {
+                    if (err) {
+                    } else if (plist) {
+                        this.handlePlistNotification(plist);
+                    } else {
+                    }
```
Comment (Fatme) [inline mobile/ios/ios-proxy-services.ts]: Can we simplify this to `if(plist) { ... }`

File: mobile/ios/ios-device.ts [hunk 37]
```diff
@@ -320,11 +327,130 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+	private startAppWithDebugger(packageFile: string, packageName: string): void {
+        this.deploy(packageFile, packageName).wait();
+        this.attachAtStartup(packageName);
+        this.runApplication(packageName).wait();
 	}
 
     public debug(packageFile: string, packageName: string): IFuture<void> {
         return (() => {
-            this.$errors.fail({formatStr:"this will come in a future version", suppressCommandHelp: true});
+            if (options["get-port"]) {
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: This line is correct, and I do not insist on changing it. Just noting that you can use ```options.getPort``` instead for a slightly shorter code.

File: mobile/ios/ios-proxy-services.ts [hunk 38]
```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
...
+
+    private buffer: string = "";
 
 	constructor(private device: Mobile.IIOSDevice,
 		private $injector: IInjector) { }
 
-	postNotification(notificationName: string) {
-		this.plistService = this.$injector.resolve(iOSCore.PlistService, {service:  this.device.startService(MobileServices.NOTIFICATION_PROXY), format: iOSCore.CoreTypes.kCFPropertyListBinaryFormat_v1_0});
-
+	public postNotification(notificationName: string) {
```
Comment (teobugslayer) [inline mobile/ios/ios-proxy-services.ts]: Mixed tabs / spaces

File: mobile/ios/ios-proxy-services.ts [hunk 39]
```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
...
+
+    private handlePlistNotification(plist: any) {
+        if (plist.type !== "dict")
+            return;
+        var value = plist.value;
+        if (!value)
+            return;
+        var command = value["Command"];
+        var name = value["Name"]
+        if (command.type !== "string" || command.value !== "RelayNotification" || name.type !== "string")
```
Comment (Fatme) [inline mobile/ios/ios-proxy-services.ts]: Why do we need `command.value !== "RelayNotification"`?

File: mobile/ios/ios-proxy-services.ts [hunk 40]
```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
...
+        array.push(callback);
+
+        return result;
+    }
+
+    public removeObserver(name: string, callback: (name: string) => void) {
+        var array = this.observers[name];
+        if (array) {
+            var index = array.indexOf(callback);
+            if (index >= 0 && index < array.length) {
```
Comment (teobugslayer) [inline mobile/ios/ios-proxy-services.ts]: Is the second check possible to fail?

File: mobile/ios/ios-proxy-services.ts [hunk 41]
```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
...
+        var start = this.buffer.indexOf("<plist");
+        var end = this.buffer.indexOf("</plist>") + 8;
+
+        while (start >= 0 && end >= 0) {
+            var plist = this.buffer.substr(start, end);
+            this.buffer = this.buffer.substr(end);
+
+            try {
+                plistlib.loadString(plist, (err: any, plist: any) => {
+                    if (err) {
```
Comment (teobugslayer) [inline mobile/ios/ios-proxy-services.ts]: The empty error check looks suspicious. Is this intended?

File: definitions/plistlib.d.ts [hunk 42]
```diff
@@ -1,3 +1,5 @@
 declare module "plistlib" {
-	export function toString(data: any): string;
+    export function toString(data: any): string;
+    export function loadBuffer(data: NodeBuffer, callback: (err: any, plist: any) => void): void;
+    export function loadString(data: string, callback: (err: any, plist: any) => void): void;
```
Comment (teobugslayer) [inline definitions/plistlib.d.ts]: Is the "err" argument really any? Can it be Error?

File: definitions/plistlib.d.ts [hunk 43]
```diff
@@ -1,3 +1,5 @@
 declare module "plistlib" {
-	export function toString(data: any): string;
+    export function toString(data: any): string;
+    export function loadBuffer(data: NodeBuffer, callback: (err: any, plist: any) => void): void;
```
Comment (teobugslayer) [inline definitions/plistlib.d.ts]: Is the "err" argument really any? Can it be Error?

File: mobile/ios/ios-proxy-services.ts [hunk 44]
```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
...
+                    } else if (plist) {
+                        this.handlePlistNotification(plist);
+                    } else {
+                    }
+                });
+            }
+            catch (e) {
```
Comment (teobugslayer) [inline mobile/ios/ios-proxy-services.ts]: Is the empty catch clause intended?

File: mobile/ios/ios-proxy-services.ts [hunk 45]
```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
...
+        var command = value["Command"];
+        var name = value["Name"]
+        if (command.type !== "string" || command.value !== "RelayNotification" || name.type !== "string")
+            return;
+        var notification = name.value;
+        var observers = this.observers[notification];
+        if (!observers)
+            return;
+
+        observers.forEach(observer => { observer(notification); });
```
Comment (teobugslayer) [inline mobile/ios/ios-proxy-services.ts]: Do you need the { ;} incantations?

File: mobile/ios/ios-proxy-services.ts (hunk 40 above)
Comment (Fatme) [inline mobile/ios/ios-proxy-services.ts]: I think it should be `if (index !== -1)`

File: mobile/ios/ios-device.ts (hunk 28 above)
Comment (PanayotCankov) [inline mobile/ios/ios-device.ts]: They may come from a command line argument in future. 8080 is really a frequently used port but we do not time to implement that now. I prefer to keep them as they are now. Is it ok? Or I MUST change them?

File: mobile/ios/ios-device.ts (hunk 29 above)
Comment (PanayotCankov) [inline mobile/ios/ios-device.ts]: Only the debug part requires the notificationproxyClient. The whole debugging logic must be refactored and extracted in a separate class where the notificationProxy will be a field. It will increase the cohesion. I don't like adding a field for something that is used by a quarter of the class methods.

File: mobile/ios/ios-proxy-services.ts (hunk 34 above)
Comment (PanayotCankov) [inline mobile/ios/ios-proxy-services.ts]: true

File: mobile/ios/ios-proxy-services.ts (hunk 33 above)
Comment (PanayotCankov) [inline mobile/ios/ios-proxy-services.ts]: nope, I will try to listen on the underlying socket.

File: mobile/ios/ios-proxy-services.ts (hunk 36 above)
Comment (PanayotCankov) [inline mobile/ios/ios-proxy-services.ts]: What if both error and plist? I will simplify it :)

File: mobile/ios/ios-proxy-services.ts (hunk 39 above)
Comment (PanayotCankov) [inline mobile/ios/ios-proxy-services.ts]: There is (perhaps very small) probability that we may get plist with another command.

File: mobile/ios/ios-proxy-services.ts (hunk 40 above)
Comment (PanayotCankov) [inline mobile/ios/ios-proxy-services.ts]: I also thought that the web socket will raise "error" when it fails to connect but in chrome it also raises "close"... I will change it if you insist :)

File: definitions/plistlib.d.ts (hunk 42 above)
Comment (PanayotCankov) [inline definitions/plistlib.d.ts]: It is a 3rd party library. I'll have no idea until I read their implementation. I do not want to give promises I can not keep.

File: mobile/ios/ios-proxy-services.ts (hunk 45 above)
Comment (PanayotCankov) [inline mobile/ios/ios-proxy-services.ts]: nope,
I ll fix that

File: mobile/ios/ios-device.ts (hunk 28 above)
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: Add a comment that we are planning to allow changing them in the future and we have a deal.

File: definitions/plistlib.d.ts (hunk 42 above)
Comment (teobugslayer) [inline definitions/plistlib.d.ts]: fair enough, 'any' it will stay.

File: mobile/ios/ios-device.ts (hunk 7 above)
Comment (PanayotCankov) [inline mobile/ios/ios-device.ts]: We get "AppLaunching" when the app is deployed, started and the debugging server is already listening for notifications. The "attachAtStartup" will add the listener, and will invoke it after the "runApplication" starts the app.

File: mobile/ios/ios-device.ts (hunk 29 above)
Comment (Fatme) [inline mobile/ios/ios-device.ts]: :+1: for extracting debug related code to another class.

File: mobile/ios/ios-proxy-services.ts (hunk 44 above)
Comment (PanayotCankov) [inline mobile/ios/ios-proxy-services.ts]: It was intended to catch parsing Errors. But it will catch handling Errors. I'll remove it.

File: definitions/plistlib.d.ts (hunk 43 above)
Comment (PanayotCankov) [inline definitions/plistlib.d.ts]: Agreed on any.

File: mobile/ios/ios-proxy-services.ts (hunk 38 above)
Comment (PanayotCankov) [inline mobile/ios/ios-proxy-services.ts]: should be tabs now

File: mobile/ios/ios-device.ts [hunk 46]
```diff
@@ -295,7 +297,7 @@ export class IOSDevice implements Mobile.IIOSDevice {
 				};
 
 				try {
-			    	this.tryToExecuteFunction<void>(func);
+					this.tryToExecuteFunction<void>(func);
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: I found a bug in unrelated code part and committed a fix in https://github.com/telerik/mobile-cli-lib/pull/201 If my change is merged before your PR, you'll need to rebase.

File: mobile/ios/ios-device.ts [hunk 47]
```diff
@@ -322,11 +328,19 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
 	}
 
-    public debug(packageFile: string, packageName: string): IFuture<void> {
-        return (() => {
-            this.$errors.fail({formatStr:"this will come in a future version", suppressCommandHelp: true});
-        }).future<void>()();
-    }
+	public debug(packageFile: string, packageName: string): IFuture<void> {
+		return (() => {
+
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: Extra empty lines

File: mobile/ios/ios-device.ts [hunk 48]
```diff
@@ -322,11 +328,19 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+			var debugging: IOSDeviceDebugging = this.$injector.resolve(IOSDeviceDebugging, {
```
Comment (teobugslayer) [inline mobile/ios/ios-device.ts]: While using manual injector resolve is correct, it makes writing unit tests harder, because one have to perform lifetime management of the injector itself. Consider using constructor injection if possible.

Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]: I think current code is fine as it passes specific values to IOSDeviceDebugging constructor.

File: mobile/ios/ios-device.ts [hunk 49]
```diff
@@ -385,5 +398,169 @@ export class IOSDevice implements Mobile.IIOSDevice {
...
+			var socketValue = socketRef.deref();
+
+			return new net.Socket({ fd: socketValue });
+		}
+
+		return null;
+	}
+
+	/**
+	 * Converts a little endian 16 bit int number to 16 bit int big endian number.
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-device.ts]: awesome, thanks for the comment

**Comment by rosen-vladimirov** [comment]
:+1:  after rebasing and handling @teobugslayer comment for empty lines

File: definitions/mobile.d.ts [hunk 50]
```diff
@@ -19,7 +19,6 @@ declare module Mobile {
 		sync(localToDevicePaths: ILocalToDevicePathData[], appIdentifier: IAppIdentifier, liveSyncUrl: string, options: ISyncOptions): IFuture<void>;
 		debug(packageFile: string, packageName: string): IFuture<void>;
 		openDeviceLogStream(): void;
-		getInstalledApplications(): IFuture<string[]>;
```
Comment (Fatme) [inline definitions/mobile.d.ts]: Do you need to remove this?

Comment (PanayotCankov) [inline definitions/mobile.d.ts]: Yes, it was a duplicated declaration of a method. The same getInstalledApplciations() is available  at the start of the class declaration.
